Fix single threaded arc fit with bad pixels.

timothyfrankdavies opened the pull request:
## Description
I ran into an issue while testing https://github.com/PyWiFeS/pipeline/pull/32. When running a particular test case multithreaded, I got different results to running single threaded. I traced the difference to the `if multithread: ` cases in `weighted_loggauss_arc_fit`.

Both single & multithreaded cases are run using a `jobs` list. Those jobs store a line index, and running `mpfit_gauss_line` or `lsq_gauss_line` on a job returns a tuple storing `(line index, center)`.

A job is made for each arc *except* arcs with bad pixels. In that case, the center is recorded as 'nan'.

So when we go to record the center for each arc, we need to skip the indexes containing bad pixels. To do that, we just need to read the line index of the job. The multithreaded case did this, but the single threaded case did not.

I've updated the single threaded case to follow the same procedure as the multithreaded case.

## Tests
I've run the new version in single threaded & multithreaded modes, and get the same results.

Here's an example compare single threaded runs before & after the fix:
![image](https://github.com/PyWiFeS/pipeline/assets/94426704/122b5d15-1389-452b-b9f9-8636e41eca1a)

Comparing multithreaded before & single threaded after:
![image](https://github.com/PyWiFeS/pipeline/assets/94426704/0fbc79a9-772d-42d4-b590-72d65032fee8)

I've done small refactors to the multithreaded case now, they pass too.

CMartinezLombilla commented on src/pywifes/wifes_wsol.py:
```diff
@@ -299,43 +309,22 @@ def weighted_loggauss_arc_fit(subbed_arc_data,
             # MJI: with the following test, the code ran faster, but even just the Pool(cpu)
             # command slowed things down. 
             #jobs2 = [jobs[:len(jobs)//4], jobs[len(jobs)//4:2*len(jobs)//4],jobs[2*len(jobs)//4:3*len(jobs)//4],jobs[3*len(jobs)//4:]]
```
Maybe we should remove this cemented-out code as `jobs` is not a given argument in `weighted_loggauss_arc_fit`.

CMartinezLombilla commented on src/pywifes/wifes_wsol.py:
```diff
@@ -299,43 +309,22 @@ def weighted_loggauss_arc_fit(subbed_arc_data,
             # command slowed things down. 
             #jobs2 = [jobs[:len(jobs)//4], jobs[len(jobs)//4:2*len(jobs)//4],jobs[2*len(jobs)//4:3*len(jobs)//4],jobs[3*len(jobs)//4:]]
-            #results = mypool.imap_unordered(utils.lsq_gauss_line,jobs2)
+            #results = mypool.imap_unordered(utils._lsq_gauss_line,jobs2)
```
Same for this commented line. Just remove it.

timothyfrankdavies replied:
Yup, agreed. I've got more refactors in the next PR, but certainly anything I'm hitting in this MR should be cleaned up now.

CMartinezLombilla commented on src/pywifes/wifes_wsol.py:
```diff
@@ -299,43 +309,22 @@ def weighted_loggauss_arc_fit(subbed_arc_data,
-                    #print('  core done in',datetime.datetime.now()-start)
+                        line_center_pairs = mypool.imap_unordered(_lsq_gauss_line,jobs)
+
+                    fitted_centers = _get_fitted_centers(line_center_pairs, fitted_centers)
```
Please, fix the indentation in lines 317 to 321.

CMartinezLombilla commented on src/pywifes/wifes_wsol.py:
```diff
@@ -299,43 +309,22 @@ def weighted_loggauss_arc_fit(subbed_arc_data,
                     #start = datetime.datetime.now() #!!! MJI
                     if find_method =='mpfit':
-                        results = mypool.imap_unordered(mpfit_gauss_line,jobs)
+                        line_center_pairs = mypool.imap_unordered(_mpfit_gauss_line,jobs)
```
Where is `jobs` defined? Am I missing something here?

CMartinezLombilla replied:
Ok! Thank you. All good now.

CMartinezLombilla commented on src/pywifes/wifes_wsol.py:
```diff
@@ -299,43 +309,22 @@ def weighted_loggauss_arc_fit(subbed_arc_data,
+                    line_center_pairs = [_mpfit_gauss_line(job) for job in jobs]
+                else:
+                    line_center_pairs = [_lsq_gauss_line(job) for job in jobs]
+                fitted_centers = _get_fitted_centers(line_center_pairs, fitted_centers)
```
Same here for the indentation.

CMartinezLombilla reviewed (commented):
This looks good, just some minor changes and one question about the `jobs` variable to take a look at. Thank you!

timothyfrankdavies commented:
I've made a number of refactors, and thankfully they haven't complicated the diff too much. There's a couple of risky changes, but I'm retesting and will make changes as needed.

Happy to undo any of the refactors.

timothyfrankdavies commented:
> I've made a number of refactors, and thankfully they haven't complicated the diff too much. There's a couple of risky changes, but I'm retesting and will make changes as needed.
> 
> Happy to undo any of the refactors.

I've now fixed the changes and retested, I get the same results as before the refactor. However, I've only tested the mpfit single-threaded codepath, on classic & ns data. The multithreaded codepath will be retested in #32.

I think the loggauss function hasn't been maintained for some time, as it used `np` instead of `numpy` and could have crashed if used. I've gone and renamed those just in case, but it's not necessarily worth it.

The refactor diff in github isn't quite as clean as one I viewed in vscode. Again, happy to undo the refactor if you think it makes it any harder to verify.
